### ISSUE TITLE
How to get pipeline parameters from a run

### ISSUE THREAD

**Issue by amueller** [openml/openml-python#745]
With @hp5200 we've been trying to analyse some runs on openml.
It would be great if there was a way to get the parameter settings from a run in a way that is similar to how they are structured in sklearn.

The best I could do so far is:
```python
import openml
run = openml.runs.get_run(2081320)
setup = openml.setups.get_setup(run.setup_id)
setup.parameters
```

These include the parameters of the sub-flows. Say I'm only interested in the last step of the pipeline. I have no idea what the order of the subflows is.  I also looked at ``flow.components``, but that is also not ordered in the way I would expect.

It's possible to get the ``steps`` of the ``Pipeline``, however they are a string:

```python
steps = [s for s in setup.parameters.values() if s.parameter_name == "steps"][0]
steps.value
```
So that's sort of a dead end unless we want to do eval or string parsing.

Somehow the pipeline is reconstructed from the flow so I'm sure the information is in there somewhere, but I haven't been able to find it. Any input welcome.

**Comment by amueller** [comment]
I feel like this is weird, but is this the recommended way to do it?

```python
flow.components[json.loads(flow.parameters['steps'])[-1]['value']['step_name']]
```

**Comment by mfeurer** [comment]
> It would be great if there was a way to get the parameter settings from a run in a way that is similar to how they are structured in sklearn

I agree.

> It's possible to get the `steps` of the Pipeline, however they are a string:

OpenML does not support lists as types for parameter values.

> Is this really how you do it?

I currently can't think of a better way, but @janvanrijn once added a function `get_structure` to the flow. Maybe it can be used to simplify this. 

Sorry, we never thought of this. I'm wondering whether that's a function worth adding to the OpenML-Python interface? @PGijsbers @janvanrijn

**Comment by PGijsbers** [comment]
I'm not aware of a better generic way to do this. I think it makes sense to add an easier way to retrieve parameters by subcomponent of the linked flow/group parameters by subflow.

**Comment by amueller** [comment]
I saw ``get_structure`` but I did not understand the docstring. Maybe an example would be helpful.

**Comment by amueller** [comment]
It would be really great if there was some documentation on how to use OpenML for meta-learning...
I feel like even very simple analysis is quite hard to figure out.

> OpenML does not support lists as types for parameter values.

You mean the OpenML schema. But that doesn't mean the OpenML python interface doesn't, right?
I find it weird that the python API has JSON strings stored in it.

**Comment by mfeurer** [comment]
> I saw get_structure but I did not understand the docstring. Maybe an example would be helpful.

@janvanrijn 

> It would be really great if there was some documentation on how to use OpenML for meta-learning. I feel like even very simple analysis is quite hard to figure out.

Yes, that would be great. Maybe we can discuss this offline at the workshop and make it a key topic? @janvanrijn has a lot of code in a contrib library. Maybe we can move over some commonly used functionality.

> You mean the OpenML schema.

Yes, but the flow object itself does not know anything about what to do with this parameter. Only the extension object knows about this. Maybe there should be a general way to interact with the extension of the flow to have access to some helper functions?

**Comment by amueller** [comment]
From what I can tell, @janvanrijn's contrib library is quite outdated, but I haven't checked in detail.

> Yes, but the flow object itself does not know anything about what to do with this parameter.

But it doesn't need to, right? The flow was created by the extension so if only the extension knows about it that's fine?

**Comment by mfeurer** [comment]
> From what I can tell, @janvanrijn's contrib library is quite outdated, but I haven't checked in detail.

I guess Jan hasn't adapted the library to the recent extension interface. But if you know what you require, and we check what's in that library, we might find a few functions which are commonly required and should be in the main package.

> The flow was created by the extension so if only the extension knows about it that's fine?

Correct. Honostly, I simply didn't think about this yet.

**Comment by amueller** [comment]
@mfeurer no worries, but that means someone could implement it :)

**Comment by amueller** [comment]
I think the real issue here is the design. The flow is basically made to mirror the OpenML schema, so going from sklearn estimator to flows mostly makes things strings.

It would be nice if we could keep the serialization to strings for the point when we want to upload to openml as much as possible.
So basically serialize_sklearn would create a pythonic representation of the flow, and only the uploading would flatten everything to a string.

This sounds like it might be a pretty annoying refactoring, though.

**Comment by mfeurer** [comment]
This basically sounds like adding in a layer of abstraction between the OpenML schema (`OpenMLFlow` class) and the scikit-learn object? Besides the refactor, I'd be afraid that we add a flow schema ourselves which we need to both define and maintain.

**Comment by amueller** [comment]
Well, the OpenMLFlow class is not really the schema, right?
The layer would be a python API to the OpenML flow, which is what ``openml-python`` is supposed to be, right?

**Comment by mfeurer** [comment]
> Well, the OpenMLFlow class is not really the schema, right?
No, it's not, but it tries to stay as close to the XML schema as possible.
> The layer would be a python API to the OpenML flow, which is what openml-python is supposed to be, right?
Could you give an example on how that would look like?

**Comment by amueller** [comment]
Hm I think I need to come up with an actual example to make this more clear.
